bigframes/core/__init__.py

```diff
@@ -49,470 +40,196 @@
 
-    @classmethod
-    def mem_expr_from_pandas(
-        cls,
-        pd_df: pandas.DataFrame,
-        session: Optional[Session],
-    ) -> ArrayValue:
-        """
-        Builds an in-memory only (SQL only) expr from a pandas dataframe.
+    @property
+    def session(self) -> Session:
+        required_session = self.node.sessions
+        from bigframes import get_global_session
 
-        Caution: If session is None, only a subset of expr functionality will
-        be available (null Session is usually not supported).
-        """
-        # We can't include any hidden columns in the ArrayValue constructor, so
-        # grab the column names before we add the hidden ordering column.
-        column_names = [str(column) for column in pd_df.columns]
-        # Make sure column names are all strings.
-        pd_df = pd_df.set_axis(column_names, axis="columns")
-        pd_df = pd_df.assign(**{ORDER_ID_COLUMN: range(len(pd_df))})
+        return self.node.sessions[0] if required_session else get_global_session()
 
-        # ibis memtable cannot handle NA, must convert to None
-        pd_df = pd_df.astype("object")  # type: ignore
-        pd_df = pd_df.where(pandas.notnull(pd_df), None)
+    def get_column_type(self, key: str) -> bigframes.dtypes.Dtype:
+        return self.compile().get_column_type(key)
 
-        # NULL type isn't valid in BigQuery, so retry with an explicit schema in these cases.
-        keys_memtable = ibis.memtable(pd_df)
-        schema = keys_memtable.schema()
-        new_schema = []
-        for column_index, column in enumerate(schema):
-            if column == ORDER_ID_COLUMN:
-                new_type: ibis_dtypes.DataType = ibis_dtypes.int64
-            else:
-                column_type = schema[column]
-                # The autodetected type might not be one we can support, such
-                # as NULL type for empty rows, so convert to a type we do
-                # support.
-                new_type = bigframes.dtypes.bigframes_dtype_to_ibis_dtype(
-                    bigframes.dtypes.ibis_dtype_to_bigframes_dtype(column_type)
-                )
-                # TODO(swast): Ibis memtable doesn't use backticks in struct
-                # field names, so spaces and other characters aren't allowed in
-                # the memtable context. Blocked by
-                # https://github.com/ibis-project/ibis/issues/7187
-                column = f"col_{column_index}"
-            new_schema.append((column, new_type))
+    def compile(self) -> compiled.CompiledArrayValue:
+        return compiled.ArrayValueCompiler().compile(self)
 
-        # must set non-null column labels. these are not the user-facing labels
-        pd_df = pd_df.set_axis(
-            [column for column, _ in new_schema],
-            axis="columns",
-        )
-        keys_memtable = ibis.memtable(pd_df, schema=ibis.schema(new_schema))
+    def shape(self) -> typing.Tuple[int, int]:
```

Comment:
Let's add something to our backlog to try and get rid of `shape` at this layer. I think it makes some caching more difficult in future.

Reply:
created task